pallets/parachain-staking/src/delegation_requests.rs

```diff
@@ -0,0 +1,317 @@
+				Self::delegator_leaves_candidate(collator.clone(), delegator.clone(), amount)?;
+				Self::deposit_event(Event::DelegationRevoked {
+					delegator: delegator.clone(),
+					candidate: collator,
+					unstaked_amount: amount,
+				});
+				if leaving {
+					<DelegatorState<T>>::remove(&delegator);
+					Self::deposit_event(Event::DelegatorLeft {
+						delegator: delegator,
+						unstaked_amount: amount,
+					});
+				} else {
+					<DelegatorState<T>>::insert(&delegator, state);
+				}
+				Ok(().into())
+			}
+			DelegationAction::Decrease(amount) => {
+				// remove from pending requests
+				Self::delegator_scheduled_requests_state_remove(&delegator, &collator);
+
+				// decrease delegation
+				for x in &mut state.delegations.0 {
+					if x.owner == collator {
+						return if x.amount > amount {
+							let amount_before: BalanceOf<T> = x.amount.into();
+							x.amount = x.amount.saturating_sub(amount);
+							state.total = state.total.saturating_sub(amount);
+							let new_total: BalanceOf<T> = state.total.into();
+							ensure!(
+								new_total >= T::MinDelegation::get(),
+								<Error<T>>::DelegationBelowMin
+							);
+							ensure!(
+								new_total >= T::MinDelegatorStk::get(),
+								<Error<T>>::DelegatorBondBelowMin
+							);
+							let mut collator_info = <CandidateInfo<T>>::get(&collator)
+								.ok_or(<Error<T>>::CandidateDNE)?;
+							T::Currency::unreserve(&delegator, amount);
+							// need to go into decrease_delegation
+							let in_top = collator_info.decrease_delegation::<T>(
+								&collator,
+								delegator.clone(),
+								amount_before,
+								amount,
+							)?;
+							<CandidateInfo<T>>::insert(&collator, collator_info);
+							let new_total_staked = <Total<T>>::get().saturating_sub(amount);
+							<Total<T>>::put(new_total_staked);
+
+							<DelegatorState<T>>::insert(&delegator, state);
+							Self::deposit_event(Event::DelegationDecreased {
+								delegator,
+								candidate: collator,
+								amount,
+								in_top,
+							});
+							Ok(().into())
+						} else {
+							// must rm entire delegation if x.amount <= less or cancel request
+							Err(<Error<T>>::DelegationBelowMin.into())
+						};
+					}
+				}
+				Err(<Error<T>>::DelegationDNE.into())
+			}
+		}
+	}
+
+	/// Adds a [ScheduledRequest] to the delegator's storage state, towards a given collator.
+	pub(crate) fn delegator_scheduled_requests_state_add(
+		delegator: DelegatorId<T>,
+		collator: CollatorId<T>,
+		request: ScheduledRequest<BalanceOf<T>>,
+	) {
+		let amount = request.action.amount();
+		let is_revoke = matches!(&request.action, &DelegationAction::Revoke(_));
+
+		<DelegatorScheduledRequests<T>>::insert(delegator.clone(), collator, request);
+		if is_revoke {
+			<DelegatorScheduledRevokeRequestCount<T>>::mutate(&delegator, |x| {
+				*x = x.saturating_add(1)
+			});
+		}
+		<DelegatorScheduledRequestDecreaseAmount<T>>::mutate(&delegator, |x| {
+			*x = x.saturating_add(amount)
+		});
```

Comment:
We `get` this storage item in the body of the calling function. Then we `mutate` here which is 1 read + 1 write. So now it is 2 reads + 1 write.

In the old code, we got the old storage item and inserted the new item so it was only 1 read + 1 write.

It's probably not a big difference cc @notlesh

Reply:
In practice this wouldn't be a problem because re-reading a storage item should read from cache and won't further impact PoV.

It looks like the benchmarking code tracks `repeat_reads` independently, so I imagine it is smart about that.

Reply:
From what I read of storage, repeated reads are cached, and so are write-then-reads.

Reply:
Sounds good then